net/portmapper/portmapper.go

```diff
@@ -0,0 +1,481 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package portmapper is a UDP port mapping client. It currently only does
+// NAT-PMP, but will likely do UPnP and perhaps PCP later.
+package portmapper
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"go4.org/mem"
+	"inet.af/netaddr"
+	"tailscale.com/net/interfaces"
+	"tailscale.com/net/netns"
+	"tailscale.com/types/logger"
+)
+
+//lint:file-ignore U1000 Ignore unused WIP code
+
+// References:
+//
+// NAT-PMP: https://tools.ietf.org/html/rfc6886
+// PCP: https://tools.ietf.org/html/rfc6887
+
+// portMapServiceTimeout is the time we wait for port mapping
+// services (UPnP, NAT-PMP, PCP) to respond before we give up and
+// decide that they're not there. Since these services are on the
+// same LAN as this machine and a single L3 hop away, we don't
+// give them much time to respond.
+const portMapServiceTimeout = 250 * time.Millisecond
+
+// Client is a port mapping client.
+type Client struct {
+	logf logger.Logf
+
+	mu       sync.Mutex // guards following, and all fields thereof
+	lastMyIP netaddr.IP
+	lastGW   netaddr.IP
+	closed   bool
+
+	localPort  uint16
+	pmpMapping *pmpMapping // non-nil if we have a PMP mapping
+}
+
+type pmpMapping struct {
+	gw netaddr.IP
+
+	// all fields guarded by Client.mu
+
+	external netaddr.IPPort
+	internal netaddr.IPPort
+	useUntil time.Time
+	epoch    uint32
+}
+
+// externalValid reports whether m.external is valid, with both its IP and Port populated.
+func (m *pmpMapping) externalValid() bool {
+	return !m.external.IP.IsZero() && m.external.Port != 0
```

Comment:
Should netaddr have an IPPort.IsZero() ?

Reply:
It does, but that's not what this wants to check: this wants to check that BOTH fields are non-zero, not either.